pkg/hnsw/vector.go

```diff
@@ -0,0 +1,119 @@
+package hnsw
+
+import (
+	"errors"
+	"math"
+)
+
+type Point []float32
+
+type Vector struct {
+	id    Id
+	point Point
```

Comment:
Yeah, so this is getting closer. Note that the only user of `point` is the euclidean distance function, so if we don't store it all together then `Vector` can be

```
type Vector struct {
  friends []*BaseQueue
}
```

which is quite a bit cleaner (in particular because you could go as far as getting rid of `Vector` all together with some clever writing.

This would require pulling out euclidean distance into a pure function instead of on the struct but I think that's a good idea.